### PR TITLE
Optimize cosine shape handling in rope.py to adapt Wan2.2

### PR DESCRIPTION
**Pull request by Li-8916**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Optimize cosine shape handling in rope.py, when the shape is [1,s,1,d], it can use rotary_position_embedding direclty
## Test Plan
verify the modification on Wan2.2 

## Test Result
1、verify the modification on vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py

add the below command:
from importlib.util import find_spec
from vllm_omni.diffusion.layers.rope import RotaryEmbedding
from vllm_omni.platforms import current_omni_platform

WanSelfAttention
init:
self.rope = RotaryEmbedding(is_neox_style=False)

forward:
if rotary_emb is not None:
            freqs_cos, freqs_sin = rotary_emb
            if find_spec("mindiesd") is not None and current_omni_platform.is_npu():
                query = self.rope(query,freqs_cos, freqs_sin)
                key = self.rope(key,freqs_cos, freqs_sin)
            else:
                query = apply_rotary_emb_wan(query, freqs_cos, freqs_sin)
                key = apply_rotary_emb_wan(key, freqs_cos, freqs_sin)

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
## Step 0: 审核门槛检查

❌ **BLOCKING: DCO 检查失败**

在继续审核之前，需要先修复 DCO (Developer Certificate of Origin) 检查失败的问题。

**如何修复：**
1. 确保你的 commit 包含 `Signed-off-by: Your Name <your.email@example.com>`
2. 可以使用 `git commit --amend --signoff` 添加签名
3. 强制推送更新后的 commit

---

## Blocker Scan

| Category            | Result                                  |
|---------------------|-----------------------------------------|
| Correctness         | ✅ PASS                                 |
| Reliability/Safety  | ✅ PASS                                 |
| Breaking Changes    | ✅ PASS                                 |
| Test Coverage       | ⚠️ ISSUES: 缺少自动化测试               |
| Documentation       | ✅ PASS                                 |
| Security            | ✅ PASS                                 |

---

## 代码分析

**变更内容：**
```python
+    if cos.shape[-1] == x.shape[-1]:
+        half_head_dim = False
+
```

**逻辑理解：**
- 当 `cos` 的最后一个维度（D）等于 `x` 的最后一个维度（headdim）时，表示输入已经是完整维度（不是 D/2）
- 此时设置 `half_head_dim = False`，避免后续不必要的维度扩展操作
- 这是一个针对 Wan2.2 模型（shape 为 [1, s, 1, d]）的优化

**Correctness 检查：** ✅
- 逻辑正确：当 cos.shape[-1] == x.shape[-1] 时，确实表示输入已经是完整维度
- 位置合理：在 `cos.dim() == 3` 处理之后，在 `interleaved` 分支之前
- 不会影响现有行为：只有当维度匹配时才触发优化，其他情况保持原有逻辑

**Breaking Changes 检查：** ✅
- 这是一个优化，不改变函数接口或默认行为
- 只影响特定 shape 的输入（Wan2.2）
- 对其他模型无影响

---

## 需要改进的地方

### 1. ❌ DCO 检查失败 (BLOCKING)

**问题：** Commit 缺少 DCO 签名

**修复方法：**
```bash
git commit --amend --signoff
git push --force
```

---

### 2. ⚠️ 缺少自动化测试 (NON-BLOCKING)

**问题：**
- PR 描述中只提供了手动验证方法
- 没有添加自动化测试来确保优化不影响正确性
- 未来修改可能破坏这个优化

**建议：**
添加单元测试验证优化逻辑：

```python
# tests/diffusion/layers/test_rope.py

def test_apply_rotary_emb_mindiesd_full_dim_optimization():
    """Test that when cos.shape[-1] == x.shape[-1], optimization is applied"""
    import torch
    from vllm_omni.diffusion.layers.rope import apply_rotary_emb_mindiesd
    
    # Test case: full dimension (D, not D/2)
    batch, seqlen, nheads, headdim = 1, 16, 8, 64
    x = torch.randn(batch, seqlen, nheads, headdim)
    cos = torch.randn(seqlen, headdim)  # Full dimension, not headdim/2
    sin = torch.randn(seqlen, headdim)
    
    # Should use optimized path (half_head_dim=False)
    output = apply_rotary_emb_mindiesd(x, cos, sin, interleaved=False)
    
    # Verify output shape
    assert output.shape == x.shape
    
def test_apply_rotary_emb_mindiesd_half_dim_backward_compat():
    """Test that half dimension case still works correctly"""
    import torch
    from vllm_omni.diffusion.layers.rope import apply_rotary_emb_mindiesd
    
    # Test case: half dimension (D/2)
    batch, seqlen, nheads, headdim = 1, 16, 8, 64
    x = torch.randn(batch, seqlen, nheads, headdim)
    cos = torch.randn(seqlen, headdim // 2)  # Half dimension
    sin = torch.randn(seqlen, headdim // 2)
    
    # Should use original path (half_head_dim=True)
    output = apply_rotary_emb_mindiesd(x, cos, sin, interleaved=False)
    
    # Verify output shape
    assert output.shape == x.shape
```

---

## PR 描述改进建议

建议在 PR 描述中补充以下信息：

**Test Environment:**
```
Test env: NPU (Ascend), MindIESD, Wan2.2 model
Commands: (provide exact commands used for verification)
Runtime: ~N min
```

**Expected Behavior:**
- Before optimization: apply_rotary_emb_mindiesd performs unnecessary dimension expansion
- After optimization: directly use rotary_position_embedding when dimensions already match
- Performance improvement: (if measured, provide latency comparison)

---

## 总结

**BLOCKERS:**
1. ❌ DCO 检查失败 - 必须修复才能合并

**NON-BLOCKING SUGGESTIONS:**
1. ⚠️ 添加自动化测试验证优化逻辑
2. 📝 完善 PR 描述（测试环境、性能对比）

**VERDICT:** REQUEST_CHANGES

修复 DCO 后，这是一个值得合并的优化。代码逻辑正确，不会引入破坏性变更。

---

**下一步：**
1. 修复 DCO 检查（使用 `git commit --amend --signoff`）
2. （可选但推荐）添加单元测试
3. （可选）提供性能对比数据
4. 重新请求审核

**Comment by Li-8916** [comment]
> ## Step 0: 审核门槛检查
> ❌ **BLOCKING: DCO 检查失败**
> 
> 在继续审核之前，需要先修复 DCO (Developer Certificate of Origin) 检查失败的问题。
> 
> **如何修复：**
> 
> 1. 确保你的 commit 包含 `Signed-off-by: Your Name <your.email@example.com>`
> 2. 可以使用 `git commit --amend --signoff` 添加签名
> 3. 强制推送更新后的 commit
> 
> ## Blocker Scan
> Category	Result
> Correctness	✅ PASS
> Reliability/Safety	✅ PASS
> Breaking Changes	✅ PASS
> Test Coverage	⚠️ ISSUES: 缺少自动化测试
> Documentation	✅ PASS
> Security	✅ PASS
> ## 代码分析
> **变更内容：**
> 
> ```python
> +    if cos.shape[-1] == x.shape[-1]:
> +        half_head_dim = False
> +
> ```
> 
> **逻辑理解：**
> 
> * 当 `cos` 的最后一个维度（D）等于 `x` 的最后一个维度（headdim）时，表示输入已经是完整维度（不是 D/2）
> * 此时设置 `half_head_dim = False`，避免后续不必要的维度扩展操作
> * 这是一个针对 Wan2.2 模型（shape 为 [1, s, 1, d]）的优化
> 
> **Correctness 检查：** ✅
> 
> * 逻辑正确：当 cos.shape[-1] == x.shape[-1] 时，确实表示输入已经是完整维度
> * 位置合理：在 `cos.dim() == 3` 处理之后，在 `interleaved` 分支之前
> * 不会影响现有行为：只有当维度匹配时才触发优化，其他情况保持原有逻辑
> 
> **Breaking Changes 检查：** ✅
> 
> * 这是一个优化，不改变函数接口或默认行为
> * 只影响特定 shape 的输入（Wan2.2）
> * 对其他模型无影响
> 
> ## 需要改进的地方
> ### 1. ❌ DCO 检查失败 (BLOCKING)
> **问题：** Commit 缺少 DCO 签名
> 
> **修复方法：**
> 
> ```shell
> git commit --amend --signoff
> git push --force
> ```
> 
> ### 2. ⚠️ 缺少自动化测试 (NON-BLOCKING)
> **问题：**
> 
> * PR 描述中只提供了手动验证方法
> * 没有添加自动化测试来确保优化不影响正确性
> * 未来修改可能破坏这个优化
> 
> **建议：** 添加单元测试验证优化逻辑：
> 
> ```python
> # tests/diffusion/layers/test_rope.py
> 
> def test_apply_rotary_emb_mindiesd_full_dim_optimization():
>     """Test that when cos.shape[-1] == x.shape[-1], optimization is applied"""
>     import torch
>     from vllm_omni.diffusion.layers.rope import apply_rotary_emb_mindiesd
>     
>     # Test case: full dimension (D, not D/2)
>     batch, seqlen, nheads, headdim = 1, 16, 8, 64
>     x = torch.randn(batch, seqlen, nheads, headdim)
>     cos = torch.randn(seqlen, headdim)  # Full dimension, not headdim/2
>     sin = torch.randn(seqlen, headdim)
>     
>     # Should use optimized path (half_head_dim=False)
>     output = apply_rotary_emb_mindiesd(x, cos, sin, interleaved=False)
>     
>     # Verify output shape
>     assert output.shape == x.shape
>     
> def test_apply_rotary_emb_mindiesd_half_dim_backward_compat():
>     """Test that half dimension case still works correctly"""
>     import torch
>     from vllm_omni.diffusion.layers.rope import apply_rotary_emb_mindiesd
>     
>     # Test case: half dimension (D/2)
>     batch, seqlen, nheads, headdim = 1, 16, 8, 64
>     x = torch.randn(batch, seqlen, nheads, headdim)
>     cos = torch.randn(seqlen, headdim // 2)  # Half dimension
>     sin = torch.randn(seqlen, headdim // 2)
>     
>     # Should use original path (half_head_dim=True)
>     output = apply_rotary_emb_mindiesd(x, cos, sin, interleaved=False)
>     
>     # Verify output shape
>     assert output.shape == x.shape
> ```
> 
> ## PR 描述改进建议
> 建议在 PR 描述中补充以下信息：
> 
> **Test Environment:**
> 
> ```
> Test env: NPU (Ascend), MindIESD, Wan2.2 model
> Commands: (provide exact commands used for verification)
> Runtime: ~N min
> ```
> 
> **Expected Behavior:**
> 
> * Before optimization: apply_rotary_emb_mindiesd performs unnecessary dimension expansion
> * After optimization: directly use rotary_position_embedding when dimensions already match
> * Performance improvement: (if measured, provide latency comparison)
> 
> ## 总结
> **BLOCKERS:**
> 
> 1. ❌ DCO 检查失败 - 必须修复才能合并
> 
> **NON-BLOCKING SUGGESTIONS:**
> 
> 1. ⚠️ 添加自动化测试验证优化逻辑
> 2. 📝 完善 PR 描述（测试环境、性能对比）
> 
> **VERDICT:** REQUEST_CHANGES
> 
> 修复 DCO 后，这是一个值得合并的优化。代码逻辑正确，不会引入破坏性变更。
> 
> **下一步：**
> 
> 1. 修复 DCO 检查（使用 `git commit --amend --signoff`）
> 2. （可选但推荐）添加单元测试
> 3. （可选）提供性能对比数据
> 4. 重新请求审核
